src/python/pants/backend/nfpm/util_rules/sandbox.py

```diff
@@ -0,0 +1,258 @@
+        remaining_targets: list[Target] = []
+
+        # NB: TransitiveTargets is AFTER target generation/expansion (so there are no target generators)
+        for tgt in transitive_targets.dependencies:
+            if tgt.has_field(NfpmContentDirDstField) or tgt.has_field(NfpmContentSymlinkDstField):
+                # NfpmContentDir and NfpmContentSymlink targets don't go in the sandbox.
+                # They're only registered in the nfpm config.
+                continue
+            elif tgt.has_field(NfpmContentDstField):
+                # an NfpmContentFile DOES need something in the sandbox
+
+                # 'source' must be either None or a non-empty string.
+                # If 'source' is None, the file comes from dependencies.
+                if tgt[NfpmContentFileSourceField].value is None:
+                    # The file must be hydrated from one of the dependencies.
+                    nfpm_content_from_dependency_targets.append(cast(NfpmContentFile, tgt))
+                    continue
+                # The file must be hydrated from the 'source' field
+                nfpm_content_from_source_targets.append(cast(NfpmContentFile, tgt))
+                continue
+
+            # This bool serves as a "continue" for the outer "for tgt" loop.
+            identified_target = False
+
+            for field_set_type in NFPM_PACKAGE_FIELD_SET_TYPES:
+                if field_set_type.is_applicable(tgt):
+                    identified_target = True
+                    # we only respect nfpm package deps for the same packager
+                    # (For example, deb targets will ignore any deps on rpm targets)
+                    if isinstance(field_set, field_set_type):
+                        nfpm_package_targets.append(cast(NfpmPackageTarget, tgt))
+                    break
+            if identified_target:
+                continue
+
+            for field_set_type in package_field_set_types:
+                if field_set_type.is_applicable(tgt):
+                    identified_target = True
+                    package_targets.append(tgt)
+                    break
+            if identified_target:
+                continue
+
+            remaining_targets.append(tgt)
+
+        return cls(
+            nfpm_content_from_dependency_targets=tuple(nfpm_content_from_dependency_targets),
+            nfpm_content_from_source_targets=tuple(nfpm_content_from_source_targets),
+            nfpm_package_targets=tuple(nfpm_package_targets),
+            package_targets=tuple(package_targets),
+            remaining_targets=tuple(remaining_targets),
+        )
+
+
+@dataclass(frozen=True)
+class NfpmContentSandboxRequest:
+    field_set: NfpmPackageFieldSet
+
+
+@dataclass(frozen=True)
+class NfpmContentSandbox:
+    digest: Digest
+
+
+@rule
+async def populate_nfpm_content_sandbox(
+    request: NfpmContentSandboxRequest, union_membership: UnionMembership
+) -> NfpmContentSandbox:
```

Comment:
This will be covered by the tests for the `package_nfpm_package` rule.

Maybe this could use tests to ensure it
- does random codegen and package gen correctly
- includes all the `nfpm_content_file(source=...)` sources correctly